### ISSUE TITLE
Resources don't have the "sync-checksum" annotation when created by flux if there is annotation with value true unquoted

### ISSUE THREAD

**Issue by primeroz** [fluxcd/flux#2091]
**Describe the bug**
Some resources are missing the "sync-checksum" annotation when created by flux.

Edit: See [comment](https://github.com/weaveworks/flux/issues/2091#issuecomment-495634216)
The problem is with an annotation where the value is the string _true_ unquoted

Flux will start to print this error in the logs 
```
caller=sync.go:150 component=cluster warning="resource to be synced has not been updated; skipping" resource=vault:configmap/vault-auth-jwt
```

An example resource with this issue is [here](https://github.com/primeroz/flux-bug-test-repo/blob/example-bug/test-cm-no-checksum/ConfigMap-vault-vault-auth-jwt.yaml)

To exclude any interaction with any of the components on my clusters i setup a simple example using [kind](https://github.com/kubernetes-sigs/kind/releases) and flux pulling from a repo where only this resource exist. 
Nothing else is running in the cluster

A full example can be found in [this repo](https://github.com/primeroz/flux-bug-test-repo/tree/example-bug)

**To Reproduce**
Steps to reproduce the behaviour:
1. fork the [repo:tag](https://github.com/primeroz/flux-bug-test-repo/tree/example-bug)
2. set the correct git-repo/branch in _./scripts/flux.sh_
3. run _./scripts/start-kind.sh_ and wait for the cluster to be up
4. run _./scripts/flux.sh_ and wait for flux to be up
5. tail the logs for flux, grab the identity public key and add it as RW deploy key to your repo
6. wait for flux to sync the _vault:configmap/vault-auth-jwt_ resource
7. Check the configmap annotation to see the _sync-checksum_ is missing
8. Check the logs to see the error

**Expected behavior**
Resource should have the _sync-checksum_ annotation

**Logs**
```
flux-7965b4c4dc-wwcqz flux ts=2019-05-24T12:07:35.879221599Z caller=loop.go:103 component=sync-loop event=refreshed url=git@github.com:primeroz/flux-bug-test-repo.git branch=mas
ter HEAD=4aa9557caec537152e0ec07da3c4f2dec2bd9d6c
flux-7965b4c4dc-wwcqz flux ts=2019-05-24T12:07:37.780231119Z caller=sync.go:476 component=cluster method=Sync cmd=apply args= count=1
flux-7965b4c4dc-wwcqz flux ts=2019-05-24T12:07:38.158416373Z caller=sync.go:542 component=cluster method=Sync cmd="kubectl apply -f -" took=378.018229ms err=null output="configm
ap/vault-auth-jwt created"
flux-7965b4c4dc-wwcqz flux ts=2019-05-24T12:07:41.615700947Z caller=sync.go:150 component=cluster warning="resource to be synced has not been updated; skipping" resource=vault:c
onfigmap/vault-auth-jwt
```

**Additional context**
Add any other context about the problem here, e.g 
- Flux version: 1.12.3
- Kubernetes version: 1.12 ( both on example KIND and real GKE cluster)
- Git provider: both gitlab and github

**Comment by 2opremio** [comment]
Fantastic bug report, thanks. I will try your repo and come back with what I found out.

**Comment by primeroz** [comment]
> Fantastic bug report
Thanks ! :) 

I have been scrapping things out of the resource to trying and figure out what causes the issue. 

Removing all the annotations from the CM actually fix the problem ... if it helps

**Comment by primeroz** [comment]
So apparently the problem is with the "true" string in [this annotation](https://github.com/primeroz/flux-bug-test-repo/blob/master/test-cm-no-checksum/ConfigMap-vault-vault-auth-jwt.yaml#L70)

When quoted, as in [here](https://github.com/primeroz/flux-bug-test-repo/blob/testing-what-breaks-it/test-cm-no-checksum/ConfigMap-vault-vault-auth-jwt.yaml#L70) it works

I will fix my resources to make sure the "true" strings are all quoted  ... still worth fixing though ?

**Comment by 2opremio** [comment]
> still worth fixing though ?

I think so, we should at least output an error or a warning about the annotation not being applied properly

**Comment by primeroz** [comment]
Great, when that fix land could you also update the "kustomize" image out of your personal docker hub ?

I actually run _2opremio/flux:generators-releasers-bb344048_ which has the same issue.
thanks!

**Comment by 2opremio** [comment]
@primeroz I will try. The priority is to get that branch merged in master :)

**Comment by 2opremio** [comment]
@primeroz Unfortunately I don't think there is much we can do about it :( . `kubectl apply` (even on version kubernetes/kubectl version 1.14) swallows the definition without quotes, which results on the resource being created without annotations

**Comment by 2opremio** [comment]
@primeroz would you mind creating an upstream bug?

**Comment by 2opremio** [comment]
Uhm, I am not running 1.14 on the server:

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

@hiddeco if you are, could you give a try to creating [this resource](https://github.com/primeroz/flux-bug-test-repo/blob/example-bug/test-cm-no-checksum/ConfigMap-vault-vault-auth-jwt.yaml) to see what happens with the annotations?

**Comment by hiddeco** [comment]
They are being removed from the resource.

```
~
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"archive", BuildDate:"2019-05-20T23:31:52Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:02:58Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}

❯ kubectl create ns vault
namespace/vault created

~
❯ kubectl apply -f https://raw.githubusercontent.com/primeroz/flux-bug-test-repo/example-bug/test-cm-no-checksum/ConfigMap-vault-vault-auth-jwt.yaml
configmap/vault-auth-jwt created

~
❯ kubectl get -n vault configmap/vault-auth-jwt -o yaml
apiVersion: v1
data:
  vault-config.yml: |-
    auth:
      - type: jwt
        path: jwt
        config:
          oidc_discovery_url: https://dex.svc.vault.k8s-test.mintel.com
          oidc_discovery_ca_pem: |
            -----BEGIN CERTIFICATE-----
            MIIFATCCAumgAwIBAgIRAKc9ZKBASymy5TLOEp57N98wDQYJKoZIhvcNAQELBQAw
            GjEYMBYGA1UEAwwPRmFrZSBMRSBSb290IFgxMB4XDTE2MDMyMzIyNTM0NloXDTM2
            MDMyMzIyNTM0NlowGjEYMBYGA1UEAwwPRmFrZSBMRSBSb290IFgxMIICIjANBgkq
            hkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA+pYHvQw5iU3v2b3iNuYNKYgsWD6KU7aJ
            diddtZQxSWYzUI3U0I1UsRPTxnhTifs/M9NW4ZlV13ZfB7APwC8oqKOIiwo7IwlP
            xg0VKgyz+kT8RJfYr66PPIYP0fpTeu42LpMJ+CKo9sbpgVNDZN2z/qiXrRNX/VtG
            TkPV7a44fZ5bHHVruAxvDnylpQxJobtCBWlJSsbIRGFHMc2z88eUz9NmIOWUKGGj
            EmP76x8OfRHpIpuxRSCjn0+i9+hR2siIOpcMOGd+40uVJxbRRP5ZXnUFa2fF5FWd
            O0u0RPI8HON0ovhrwPJY+4eWKkQzyC611oLPYGQ4EbifRsTsCxUZqyUuStGyp8oa
            aoSKfF6X0+KzGgwwnrjRTUpIl19A92KR0Noo6h622OX+4sZiO/JQdkuX5w/HupK0
            A0M0WSMCvU6GOhjGotmh2VTEJwHHY4+TUk0iQYRtv1crONklyZoAQPD76hCrC8Cr
            IbgsZLfTMC8TWUoMbyUDgvgYkHKMoPm0VGVVuwpRKJxv7+2wXO+pivrrUl2Q9fPe
            Kk055nJLMV9yPUdig8othUKrRfSxli946AEV1eEOhxddfEwBE3Lt2xn0hhiIedbb
            Ftf/5kEWFZkXyUmMJK8Ra76Kus2ABueUVEcZ48hrRr1Hf1N9n59VbTUaXgeiZA50
            qXf2bymE6F8CAwEAAaNCMEAwDgYDVR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMB
            Af8wHQYDVR0OBBYEFMEmdKSKRKDm+iAo2FwjmkWIGHngMA0GCSqGSIb3DQEBCwUA
            A4ICAQBCPw74M9X/Xx04K1VAES3ypgQYH5bf9FXVDrwhRFSVckria/7dMzoF5wln
            uq9NGsjkkkDg17AohcQdr8alH4LvPdxpKr3BjpvEcmbqF8xH+MbbeUEnmbSfLI8H
            sefuhXF9AF/9iYvpVNC8FmJ0OhiVv13VgMQw0CRKkbtjZBf8xaEhq/YqxWVsgOjm
            dm5CAQ2X0aX7502x8wYRgMnZhA5goC1zVWBVAi8yhhmlhhoDUfg17cXkmaJC5pDd
            oenZ9NVhW8eDb03MFCrWNvIh89DDeCGWuWfDltDq0n3owyL0IeSn7RfpSclpxVmV
            /53jkYjwIgxIG7Gsv0LKMbsf6QdBcTjhvfZyMIpBRkTe3zuHd2feKzY9lEkbRvRQ
            zbh4Ps5YBnG6CKJPTbe2hfi3nhnw/MyEmF3zb0hzvLWNrR9XW3ibb2oL3424XOwc
            VjrTSCLzO9Rv6s5wi03qoWvKAQQAElqTYRHhynJ3w6wuvKYF5zcZF3MDnrVGLbh1
            Q9ePRFBCiXOQ6wPLoUhrrbZ8LpFUFYDXHMtYM7P9sc9IAWoONXREJaO08zgFtMp4
            8iyIYUyQAbsvx8oD2M8kRvrIRSrRJSl6L957b4AFiLIQ/GgV2curs0jje7Edx34c
            idWw1VrejtwclobqNMVtG3EiPUIpJGpbMcJgbiLSmKkrvQtGng==
            -----END CERTIFICATE-----
          bound_issuer: https://dex.svc.vault.k8s-test.mintel.com
          default_role: default
        roles:
          - name: mintel
            role_type: jwt
            bound_audiences:
              - vault
            user_claim: email
            groups_claim: "groups"
            policies: ""
            ttl: 10m
            max_ttl: 24h
    groups:
      - name: admin
        policies:
          - admin_access
          - gcp_admin_access
        metadata:
          privileged: "true"
        type: external
    group-aliases:
      - name: k8s-admin
        mountpath: jwt
        group: admin
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"vault-config.yml":"auth:\n  - type: jwt\n    path: jwt\n    config:\n      oidc_discovery_url: https://dex.svc.vault.k8s-test.mintel.com\n      oidc_discovery_ca_pem: |\n        -----BEGIN CERTIFICATE-----\n        MIIFATCCAumgAwIBAgIRAKc9ZKBASymy5TLOEp57N98wDQYJKoZIhvcNAQELBQAw\n        GjEYMBYGA1UEAwwPRmFrZSBMRSBSb290IFgxMB4XDTE2MDMyMzIyNTM0NloXDTM2\n        MDMyMzIyNTM0NlowGjEYMBYGA1UEAwwPRmFrZSBMRSBSb290IFgxMIICIjANBgkq\n        hkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA+pYHvQw5iU3v2b3iNuYNKYgsWD6KU7aJ\n        diddtZQxSWYzUI3U0I1UsRPTxnhTifs/M9NW4ZlV13ZfB7APwC8oqKOIiwo7IwlP\n        xg0VKgyz+kT8RJfYr66PPIYP0fpTeu42LpMJ+CKo9sbpgVNDZN2z/qiXrRNX/VtG\n        TkPV7a44fZ5bHHVruAxvDnylpQxJobtCBWlJSsbIRGFHMc2z88eUz9NmIOWUKGGj\n        EmP76x8OfRHpIpuxRSCjn0+i9+hR2siIOpcMOGd+40uVJxbRRP5ZXnUFa2fF5FWd\n        O0u0RPI8HON0ovhrwPJY+4eWKkQzyC611oLPYGQ4EbifRsTsCxUZqyUuStGyp8oa\n        aoSKfF6X0+KzGgwwnrjRTUpIl19A92KR0Noo6h622OX+4sZiO/JQdkuX5w/HupK0\n        A0M0WSMCvU6GOhjGotmh2VTEJwHHY4+TUk0iQYRtv1crONklyZoAQPD76hCrC8Cr\n        IbgsZLfTMC8TWUoMbyUDgvgYkHKMoPm0VGVVuwpRKJxv7+2wXO+pivrrUl2Q9fPe\n        Kk055nJLMV9yPUdig8othUKrRfSxli946AEV1eEOhxddfEwBE3Lt2xn0hhiIedbb\n        Ftf/5kEWFZkXyUmMJK8Ra76Kus2ABueUVEcZ48hrRr1Hf1N9n59VbTUaXgeiZA50\n        qXf2bymE6F8CAwEAAaNCMEAwDgYDVR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMB\n        Af8wHQYDVR0OBBYEFMEmdKSKRKDm+iAo2FwjmkWIGHngMA0GCSqGSIb3DQEBCwUA\n        A4ICAQBCPw74M9X/Xx04K1VAES3ypgQYH5bf9FXVDrwhRFSVckria/7dMzoF5wln\n        uq9NGsjkkkDg17AohcQdr8alH4LvPdxpKr3BjpvEcmbqF8xH+MbbeUEnmbSfLI8H\n        sefuhXF9AF/9iYvpVNC8FmJ0OhiVv13VgMQw0CRKkbtjZBf8xaEhq/YqxWVsgOjm\n        dm5CAQ2X0aX7502x8wYRgMnZhA5goC1zVWBVAi8yhhmlhhoDUfg17cXkmaJC5pDd\n        oenZ9NVhW8eDb03MFCrWNvIh89DDeCGWuWfDltDq0n3owyL0IeSn7RfpSclpxVmV\n        /53jkYjwIgxIG7Gsv0LKMbsf6QdBcTjhvfZyMIpBRkTe3zuHd2feKzY9lEkbRvRQ\n        zbh4Ps5YBnG6CKJPTbe2hfi3nhnw/MyEmF3zb0hzvLWNrR9XW3ibb2oL3424XOwc\n        VjrTSCLzO9Rv6s5wi03qoWvKAQQAElqTYRHhynJ3w6wuvKYF5zcZF3MDnrVGLbh1\n        Q9ePRFBCiXOQ6wPLoUhrrbZ8LpFUFYDXHMtYM7P9sc9IAWoONXREJaO08zgFtMp4\n        8iyIYUyQAbsvx8oD2M8kRvrIRSrRJSl6L957b4AFiLIQ/GgV2curs0jje7Edx34c\n        idWw1VrejtwclobqNMVtG3EiPUIpJGpbMcJgbiLSmKkrvQtGng==\n        -----END CERTIFICATE-----\n      bound_issuer: https://dex.svc.vault.k8s-test.mintel.com\n      default_role: default\n    roles:\n      - name: mintel\n        role_type: jwt\n        bound_audiences:\n          - vault\n        user_claim: email\n        groups_claim: \"groups\"\n        policies: \"\"\n        ttl: 10m\n        max_ttl: 24h\ngroups:\n  - name: admin\n    policies:\n      - admin_access\n      - gcp_admin_access\n    metadata:\n      privileged: \"true\"\n    type: external\ngroup-aliases:\n  - name: k8s-admin\n    mountpath: jwt\n    group: admin"},"kind":"ConfigMap","metadata":{"annotations":{},"labels":{"app":"vault-configurator","vault_cr":"vault"},"name":"vault-auth-jwt","namespace":"vault"}}
  creationTimestamp: "2019-05-24T14:15:37Z"
  labels:
    app: vault-configurator
    vault_cr: vault
  name: vault-auth-jwt
  namespace: vault
  resourceVersion: "114285"
  selfLink: /api/v1/namespaces/vault/configmaps/vault-auth-jwt
  uid: 67319c93-7e2e-11e9-9a3e-7c946f4d4b56
```

**Comment by primeroz** [comment]
@2opremio sure i can create the bug upstream , so you reckon is a "kubectl" bug ? 

I ask because they have a milion repos , if that is the case i will open here: https://github.com/kubernetes/kubectl/issues

**Comment by 2opremio** [comment]
I am not sure who does the validation, `kubectl` or the API server, but I would report it as a `kubectl` bug to start with.

Maybe the issue has already been reported ...

**Comment by hiddeco** [comment]
This should have been fixed with https://github.com/kubernetes/kubernetes/commit/c42eceb1365cfee18ebf4e09083ce112abc95b3a, available since `1.17.0`.
